contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,213 @@
+    function _postOp(PostOpMode mode, bytes calldata context, uint256 actualGasCost) internal override {
+        if (mode == PostOpMode.postOpReverted) {
+            return;
+        }
+        unchecked {
+            uint256 priceMarkup = tokenPaymasterConfig.priceMarkup;
+            uint256 preCharge = uint256(bytes32(context[0 : 32]));
+            uint256 maxFeePerGas = uint256(bytes32(context[32 : 64]));
+            address userOpSender = address(bytes20(context[64 : 84]));
+            if (mode == PostOpMode.postOpReverted) {
+                emit PostOpReverted(userOpSender, preCharge);
+                // Do nothing here to not revert the whole bundle and harm reputation
+                return;
+            }
+            uint256 _cachedPrice = updateCachedPrice(false);
+        // note: as price is in ether-per-token and we want more tokens increasing it means dividing it by markup
+            uint256 cachedPriceWithMarkup = _cachedPrice * PRICE_DENOMINATOR / priceMarkup;
+        // Refund tokens based on actual gas cost
+            uint256 actualChargeNative = actualGasCost + REFUND_POSTOP_COST * maxFeePerGas;
+            uint256 actualTokenNeeded = weiToToken(actualChargeNative, cachedPriceWithMarkup);
+            if (preCharge > actualTokenNeeded) {
+                // If the initially provided token amount is greater than the actual amount needed, refund the difference
+                SafeERC20.safeTransfer(
+                    token,
+                    userOpSender,
+                    preCharge - actualTokenNeeded
+                );
+            } else if (preCharge < actualTokenNeeded) {
+                // Attempt to cover Paymaster's gas expenses by withdrawing the 'overdraft' from the client
+                // If the transfer reverts also revert the 'postOp' to remove the incentive to cheat
+                SafeERC20.safeTransferFrom(
+                    token,
+                    userOpSender,
+                    address(this),
+                    actualTokenNeeded - preCharge
+                );
+            }
+
+            emit UserOperationSponsored(userOpSender, actualTokenNeeded, actualGasCost, cachedPrice);
+            refillEntryPointDeposit(_cachedPrice);
+        }
     }
 
-    // when constructing an account, validate constructor code and parameters
-    // we trust our factory (and that it doesn't have any other public methods)
-    function _validateConstructor(UserOperation calldata userOp) internal virtual view {
-        address factory = address(bytes20(userOp.initCode[0 : 20]));
-        require(factory == theFactory, "TokenPaymaster: wrong account factory");
+    /// @notice If necessary this function uses this Paymaster's token balance to refill the deposit on EntryPoint
+    function refillEntryPointDeposit(uint256 _cachedPrice) private {
+        uint256 currentEntryPointBalance = entryPoint.balanceOf(address(this));
```

Comment:
STATICCALL to entrypoint is not allowed
```
Limitation on “CALL” opcodes (CALL, DELEGATECALL, CALLCODE, STATICCALL):
must not use value (except from account to the entrypoint)
must not revert with out-of-gas
destination address must have code (EXTCODESIZE>0)
cannot call EntryPoint’s methods, except depositFor (to avoid recursion)
```